# Capstone Project: Building a Simple Math Chatbot (Part 2)

In Part 2, we perform the backend processing for the chatbot.

![Retrieval Augmented Generation(RAG)](https://miro.medium.com/v2/resize:fit:720/format:webp/1*UyhiO87T-hejRhqI7EwvgA.png)


## Contents
1. [Import libraries, API and set filepath](#Import-libraries,-API-and-set-filepath)
2. [Load the data](#Load-the-Data)
3. [Build index](#Build-Index)
4. [Train generation](#Train-generation)
5. [Eval generation](#Eval-generation)
6. [Initial evaluation](#Initial-evaluation)
7. [Using GPT-4 to Generate Training Data](#Using-GPT-4-to-Generate-Training-Data)
8. [Create Fine Tuned Engine](#Create-Fine-Tuned-Engine)
9. [Evaluating Fine Tuned Engine](#Evaluating-Fine-Tuned-Engine)
10. [Exploring Differences](#Exploring-Differences)

## Import libraries, API and set filepath

In [1]:
# %pip install llama-index==0.8.64 pypdf sentence-transformers ragas openai
# Operating System Interface
import os

# Load Environment Variables
from dotenv import load_dotenv, find_dotenv

# Random Number Operations
import random

# Machine Learning and AI
import openai
from llama_index.llms import OpenAI

# Data Handling and Datasets
from datasets import Dataset

# Document and Vector Store Indexing
from llama_index import Document, GPTVectorStoreIndex, ServiceContext
from llama_index import VectorStoreIndex

# Web Scraping and Directory Reading
from llama_index.readers import BeautifulSoupWebReader, SimpleDirectoryReader

# Evaluation and Dataset Generation
from llama_index.evaluation import DatasetGenerator

# Callbacks and Fine-Tuning Handlers for AI Models
from llama_index.callbacks import OpenAIFineTuningHandler, CallbackManager

# Evaluation Metrics and Utilities
from ragas import evaluate
from ragas.metrics import answer_relevancy, faithfulness

# Display Utilities for Notebooks
from llama_index.response.notebook_utils import display_response



/opt/homebrew/lib/python3.11/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


In [3]:
# load api key
_ = load_dotenv(find_dotenv()) # read local .env file
openai.api_key = os.environ['OPENAI_API_KEY']

In [7]:
# set filepath to my data directory 

current_dir = os.getcwd()
data_dir = os.path.join(current_dir, "./data")


## Load the data

According to [LlamaIndex's documentation](https://gpt-index.readthedocs.io/en/latest/examples/data_connectors/simple_directory_reader.html), the `SimpleDirectoryReader` is the most commonly used data connector that just works. Simply pass in a input directory or a list of files. It will select the best file reader based on the file extensions. 

In this use case here, there are PDFs and html pages from the latest release from HPB and MOH, which are not included in gpt-3.5-turbo's pretraining of up to Sep 2021.

In [8]:
filename_fn = lambda filename: {'file_name': filename}
pdfhtml_docs = SimpleDirectoryReader(input_dir=data_dir, exclude_hidden=True, file_metadata=filename_fn).load_data()
print([x.doc_id for x in pdfhtml_docs])
print(f"Loaded {len(pdfhtml_docs)} docs")

['66fe35d1-1297-48f1-bf9e-7b7dd664a38b', 'f095173a-7331-4c88-be1d-dfd5d36ba3a2', '64f3556e-c116-487d-8702-f892c0597d1b', 'cccebb0e-2008-4cd6-a329-044cb84ee721', '0df45f31-4473-4a72-8815-6dd849ce3de7', 'a92023d2-ddd7-4776-ba7c-e42db8aba7b9', '55860270-c5dd-45e7-a5cb-9d4490e9d5ac', '5239d46e-f76c-4eb1-8879-6ec51fd62094', 'efcac31f-ada9-480d-a9e8-cc2bd9e5ffc5', 'eab4aefa-6568-4ded-9e2f-2bb84a16901e', 'f40d5444-baa9-4ba3-8c6c-f55dd95fd365', '015e6724-8277-4094-adcb-6af9a1d12b03', '7ee6f6af-e11e-4f67-950f-1cd5c3c0b65c', '04bf9fba-cc3f-4bd6-b82d-2d8ae2f057cd', '2779b428-3d90-4d8e-80d5-6da2b997a130', 'fc904f2d-6e38-4f88-99d7-1bef264aebd2', '193a9b80-950e-4e07-a5c4-64c77e319d62', 'a39e88e4-1b8c-44e8-bf86-582c18f83f1c', 'd9dbe155-6bfe-4ea0-b6cd-ecef4b2a91dc', '2ee4c90e-b1ed-4341-a5f7-8c19f0bc3796', '05cee5c5-808c-4ebb-be01-0550af732b2e', 'fa2e7745-362e-4d4f-a4df-dec2264dab2c', '3ab87c95-9bfd-4e39-ac7c-d4329de6f264', '5355a144-08eb-452b-8eb4-bdcc7b106cee', 'dfd94765-f060-4dbe-8f40-6928c7f6d181',

## Build index

With all the data loaded, we can construct the index for the chatbot. There are 4 types of indexing: Summary index, VectorStore Index, Tree Index and Keyword Table Index. Here we are using VectorStore Index, which is also one of the most common types of indexing.

In [9]:
# for more info on service context, refer to 
# https://gpt-index.readthedocs.io/en/latest/core_modules/supporting_modules/service_context.html
service_context = ServiceContext.from_defaults(
    llm=OpenAI(model="gpt-3.5-turbo", temperature=0) # degree of randomness from 0 to 1. 
)
docs = pdfhtml_docs
index = GPTVectorStoreIndex.from_documents(documents=docs, service_context=service_context)

In [8]:
# saving the output as a vector store so that we can refer to this 
# instead of running the embedding model above again

index.storage_context.persist(persist_dir="./data/index.vecstore")


 We will be fine tuning the GPT model.

Step 1: Generate a training and evaluation dataset using GPT-4.  
Step 2: Use GPT-3.5-turbo on the evaluation questions to get our baseline performance.  
Step 3: Generate another set of training dataset using GPT-4.  
Step 4: Fine tune the GPT-3.5-turbo-0613 model on the openai website using the training dataset from Step 3.  
Step 5: Use the fine tuned model to evaluate the evaluation questions.  

## Train generation

In [203]:
# Shuffle the documents

random.seed(42)
random.shuffle(docs)

gpt_context = ServiceContext.from_defaults(
    llm=OpenAI(model="gpt-4", temperature=0)
)

In [237]:
question_gen_query=(
    "Write a primary school mathematics multiple choice question in the following form:\
    Question; A) option1 ; B) option2; C) option3 (correct); D) option4."
    )
# find out more about question generation from 
# https://gpt-index.readthedocs.io/en/latest/examples/evaluation/QuestionGeneration.html

dataset_generator = DatasetGenerator.from_documents(
    docs[:50],
    question_gen_query=question_gen_query,
    service_context=gpt_context,
)

In [3]:
import nest_asyncio
nest_asyncio.apply()


In [206]:
questions = dataset_generator.generate_questions_from_nodes(num=40)
print("Generated ", len(questions), " questions")


Generated  40  questions


In [207]:
with open("train_questions.txt", "w") as f:
    for question in questions:
        f.write(question + "\n")
        print(question)

Question: Simplify the expression 12x - 4x + 3 - 2x + 1. A) 6x + 4 B) 10x + 2 C) 6x + 2 (correct) D) 10x + 4.
What is the sum of the first two common multiples of 4 and 6? A) 24 ; B) 60; C) 36 (correct); D) 48.
What is the result of multiplying 3 260 by 100? A) 3 260 ; B) 32 600; C) 326 000 (correct); D) 3 260 000.
What is the value of 10x+5-3x when x=6? A) 45 B) 55 C) 65 (correct) D) 75
In the number 392 504, the digit 9 is in which place? A) hundreds; B) ten thousands; C) thousands (correct); D) hundred thousands.
How many hundredths are there in 0.87? A) 0.08; B) 0.8; C) 87 (correct); D) 80.
What is the numeral representation of "five million, nine thousand and sixty"? A) 5 009 006; B) 5 009 060 (correct); C) 5090 006; D) 5090 060.
Question: What is the decimal representation of 3 ones, 4 tenths, and 7 thousandths? A) 3.47 ; B) 3.407 (correct); C) 4.37; D) 3.74.
What is the area of a semicircle with a diameter of 84 cm? (Take pi = 22/7); A) 264 cm²; B) 528 cm²; C) 2772 cm² (correct)

In [208]:
input_file_path = 'train_questions.txt'
output_file_path = 'modified_train_questions.txt'

def postprocess(input_file_path, output_file_path):
    with open(input_file_path, 'r') as file:
        modified_lines = [line.replace("Question:", "").strip() for line in file]

    with open(output_file_path, 'w') as new_file:
        for line in modified_lines:
            new_file.write(line + '\n')

Screenshot of modified train questions:
![](../streamlit/images/modtrain.png)

## Eval generation

In [209]:
dataset_generator = DatasetGenerator.from_documents(
    docs[
        50:
    ],
    question_gen_query=question_gen_query,
    service_context=gpt_context,
)

In [210]:
questions = dataset_generator.generate_questions_from_nodes(num=40)
print("Generated ", len(questions), " questions")

Generated  30  questions


Screenshot of evaluation questions:
![](../streamlit/images/eval.png)

In [211]:
with open("eval_questions.txt", "w") as f:
    for question in questions:
        f.write(question + "\n")

In [212]:
input_file_path = 'eval_questions.txt'
output_file_path = 'modified_eval_questions.txt'

postprocess(input_file_path, output_file_path)

Screenshot of modified evaluation questions:
![](../streamlit/images/modeval.png)

In [213]:
print("Total number of documents:", len(docs))

Total number of documents: 80


## Initial Evaluation

For this evaluation with GPT-3.5-turbo Query Engine, we will be using the [`ragas` evaluation library](https://github.com/explodinggradients/ragas).

For this notebook, we will be using the following two metrics:

- `answer_relevancy` - This measures how relevant is the generated answer to the prompt. If the generated answer is incomplete or contains redundant information the score will be low. This is quantified by working out the chance of an LLM generating the given question using the generated answer. Values range (0,1), higher the better.  
- `faithfulness` - This measures the factual consistency of the generated answer against the given context. This is done using a multi step paradigm that includes creation of statements from the generated answer followed by verifying each of these statements against the context. The answer is scaled to (0,1) range. Higher the better.

In [214]:
questions = []
with open("modified_eval_questions.txt", "r") as f:
    for line in f:
        questions.append(line.strip())

In [215]:
# limit the context window to 2048 tokens so that refine is used
gpt_context = ServiceContext.from_defaults(
    llm=OpenAI(model="gpt-3.5-turbo", temperature=0), context_window=2048
)

index = VectorStoreIndex.from_documents(docs, service_context=gpt_context)

query_engine = index.as_query_engine(similarity_top_k=2)

In [216]:
contexts = []
answers = []

for question in questions:
    response = query_engine.query(question)
    contexts.append([x.node.get_content() for x in response.source_nodes])
    answers.append(str(response))

In [217]:
questions[:10]

['What is the remainder when 3908 is divided by 7? A) 2 ; B) 3 (correct); C) 5; D) 6',
 'What is the simplified form of the expression 6 + 8y + 7 - 5y? A) 1+ 13y ; B) 13 + 18y; C) 13 + 3y (correct); D) 13 - 3y.',
 'What is the approximate weight of a completely filled can of soft drink? A) 30g; B) 300g (correct); C) 3kg; D) 30kg.',
 'What is the product of 353 and 19? A) 334 ; B) 372; C) 6770 (correct); D) 6707.',
 'Mr. Lee exchanged a $5 note for 10 coins. All the coins had the same value. What was the value of each coin?; A) 10 cents; B) 20 cents; C) 50 cents (correct); D) 1 dollar.',
 'Marilyn bought 20 packets of tissue paper for $4. How much did each packet of tissue paper cost? A) $0.50 ; B) $0.20; C) $0.05 (correct); D) $0.02.',
 'In the ratio 5:__ = 100:120, what is the missing number? A) 10 B) 15 C) 25 (correct) D) 30',
 'How many thousands are there in 4 500 000?; A) 45 ; B) 450; C) 4 500 (correct); D) 45 000.',
 '700 305 is ____ more than 680 305. A) 200 tens B) 2 thousands 

In [218]:
ds = Dataset.from_dict(
    {
        "question": questions,
        "answer": answers,
        "contexts": contexts,
    }
)

result = evaluate(ds, [answer_relevancy, faithfulness])
print(result)

evaluating with [answer_relevancy]


100%|██████████| 2/2 [00:55<00:00, 27.95s/it]
/opt/homebrew/lib/python3.11/site-packages/datasets/table.py:1395: FutureWarning: promote has been superseded by mode='default'.
  block_group = [InMemoryTable(cls._concat_blocks(list(block_group), axis=axis))]
/opt/homebrew/lib/python3.11/site-packages/datasets/table.py:1421: FutureWarning: promote has been superseded by mode='default'.
  table = cls._concat_blocks(blocks, axis=0)


evaluating with [faithfulness]


100%|██████████| 2/2 [01:35<00:00, 47.83s/it]
/opt/homebrew/lib/python3.11/site-packages/datasets/table.py:1395: FutureWarning: promote has been superseded by mode='default'.
  block_group = [InMemoryTable(cls._concat_blocks(list(block_group), axis=axis))]
/opt/homebrew/lib/python3.11/site-packages/datasets/table.py:1421: FutureWarning: promote has been superseded by mode='default'.
  table = cls._concat_blocks(blocks, axis=0)


{'ragas_score': 0.7685, 'answer_relevancy': 0.8517, 'faithfulness': 0.7000}


## Using GPT-4 to Generate Training Data

Here, we use GPT-4 and the `OpenAIFineTuningHandler` to collect data that we want to train on.

In [219]:
finetuning_handler = OpenAIFineTuningHandler()
callback_manager = CallbackManager([finetuning_handler])

gpt_4_context = ServiceContext.from_defaults(
    llm=OpenAI(model="gpt-4", temperature=0),
    context_window=2048,  # limit the context window artifically to test refine process
    callback_manager=callback_manager,
)

In [220]:
questions = []
with open("modified_train_questions.txt", "r") as f:
    for line in f:
        questions.append(line.strip())

In [221]:
from llama_index import VectorStoreIndex

index = VectorStoreIndex.from_documents(docs, service_context=gpt_4_context)

query_engine = index.as_query_engine(similarity_top_k=2)

In [222]:
for question in questions:
    response = query_engine.query(question)

## Create Fine Tuned Engine



In [223]:
finetuning_handler.save_finetuning_events("finetuning_events.jsonl")

Wrote 40 examples to finetuning_events.jsonl


![Alt text](../streamlit/images/ftmodel.png)


## Evaluating Fine Tuned Engine

After some time, your model will be done training!

The next step is running our fine-tuned model on our eval dataset again to measure any performance increase.

In [224]:
questions = []
with open("eval_questions.txt", "r") as f:
    for line in f:
        questions.append(line.strip())

In [225]:
ft_context = ServiceContext.from_defaults(
    llm=OpenAI(model="ft:gpt-3.5-turbo-0613:personal::8Qufx558",temperature=0), context_window=2048
)
index = VectorStoreIndex.from_documents(docs, service_context=ft_context)

query_engine = index.as_query_engine(similarity_top_k=2)

In [226]:
contexts = []
answers = []

for question in questions:
    response = query_engine.query(question)
    contexts.append([x.node.get_content() for x in response.source_nodes])
    answers.append(str(response))

In [227]:
ds = Dataset.from_dict(
    {
        "question": questions,
        "answer": answers,
        "contexts": contexts,
    }
)

result = evaluate(ds, [answer_relevancy, faithfulness])
print(result)

evaluating with [answer_relevancy]


100%|██████████| 2/2 [00:53<00:00, 26.90s/it]
/opt/homebrew/lib/python3.11/site-packages/datasets/table.py:1395: FutureWarning: promote has been superseded by mode='default'.
  block_group = [InMemoryTable(cls._concat_blocks(list(block_group), axis=axis))]
/opt/homebrew/lib/python3.11/site-packages/datasets/table.py:1421: FutureWarning: promote has been superseded by mode='default'.
  table = cls._concat_blocks(blocks, axis=0)


evaluating with [faithfulness]


100%|██████████| 2/2 [11:27<00:00, 343.91s/it]
/opt/homebrew/lib/python3.11/site-packages/datasets/table.py:1395: FutureWarning: promote has been superseded by mode='default'.
  block_group = [InMemoryTable(cls._concat_blocks(list(block_group), axis=axis))]
/opt/homebrew/lib/python3.11/site-packages/datasets/table.py:1421: FutureWarning: promote has been superseded by mode='default'.
  table = cls._concat_blocks(blocks, axis=0)


{'ragas_score': 0.8150, 'answer_relevancy': 0.8922, 'faithfulness': 0.7500}


| Model            | RAGAS Score | Answer Relevancy   | Faithfulness     |
|:----------------:|:-----------:|:------------------:|:--------------:|
| GPT-3.5-Turbo    | 0.7685      | 0.8517             | 0.7000         |
| Finetuned        | 0.8150      | 0.8922             | 0.7500         |

## Exploring Differences

Let's quickly compare the differences in responses, to demonstrate that fine tuning did indeed change something.

In [10]:
index = VectorStoreIndex.from_documents(docs)

In [11]:
questions = []
with open("modified_eval_questions.txt", "r") as f:
    for line in f:
        questions.append(line.strip())

In [18]:
print(questions[8])

700 305 is ____ more than 680 305. A) 200 tens B) 2 thousands C) 20 thousands (correct) D) 2 ten thousands.


## Original

In [3]:
gpt_35_context = ServiceContext.from_defaults(
    llm=OpenAI(model="gpt-3.5-turbo", temperature=0.2),
    context_window=2048,  # limit the context window artifically to test refine process
)

In [12]:
query_engine = index.as_query_engine(service_context=gpt_35_context)

response = query_engine.query(questions[8])

display_response(response)

**`Final Response:`** The answer is B) 2 thousands.

In [13]:
custom_query="What is the value of 3 in 3560? A) 3 thousands B) 3 hundreds C) 3 tens D) 3 ones"
response = query_engine.query(custom_query)

display_response(response)

**`Final Response:`** 3 hundreds

## Fine-tuned

In [15]:
ft_context = ServiceContext.from_defaults(
    #llm=ft_llm,
    llm=OpenAI(model="ft:gpt-3.5-turbo-0613:personal::8Qufx558",temperature=0),
    context_window=2048,  # limit the context window artifically to test refine process
)

In [16]:
query_engine = index.as_query_engine(service_context=ft_context)

response = query_engine.query(questions[8])

display_response(response)

**`Final Response:`** The correct answer is C) 20 thousands.

In [17]:
response = query_engine.query(custom_query)

display_response(response)

**`Final Response:`** The value of 3 in 3560 is 3 thousands.

## Observation

There is indeed a difference between the two models. The fine-tuned model is more accurate in its answer.